daemon/k8s_watcher.go

```diff
@@ -932,11 +937,61 @@ func (d *Daemon) addCiliumNetworkPolicy(obj interface{}) {
 		}
 	}
 
+	var cnpns k8s.CiliumNetworkPolicyNodeStatus
 	if err != nil {
-		log.Warningf("Unable to add CiliumNetworkPolicy %s: %s", rule.Metadata.Name, err)
+		cnpns = k8s.CiliumNetworkPolicyNodeStatus{
+			OK:          false,
+			Error:       fmt.Sprintf("%s", err),
+			LastUpdated: time.Now(),
+		}
+		log.Warningf("Unable to add CiliumNetworkPolicy %s: err: '%s'. err != nil: '%t'. a nil object: '%s'", rule.Metadata.Name, err, err != nil, nil)
 	} else {
+		cnpns = k8s.CiliumNetworkPolicyNodeStatus{
+			OK:          true,
+			Error:       "OK",
+			LastUpdated: time.Now(),
+		}
 		log.Infof("Imported CiliumNetworkPolicy %s", rule.Metadata.Name)
 	}
+
+	rule.SetPolicyStatus(nodeaddress.GetName(), cnpns)
+	_, err = crdClient.Update(rule)
+	if err != nil {
+		ns := k8s.ExtractNamespace(&rule.Metadata)
+		name := rule.Metadata.GetObjectMeta().GetName()
+		log.Warningf("k8s: unable to update CNP %s/%s with status: %s, retrying...", ns, name, err)
+		go func(ns, name string, rule *k8s.CiliumNetworkPolicy) {
+			for n := 0; n < 10; n++ {
+				serverRuleStore, exists, err := ciliumRulesStore.Get(rule)
+				if !exists {
+					break
+				}
+				if err != nil {
+					log.Warningf("k8s: unable to get k8s CNP %s/%s from local cache: %s", ns, name, err)
+					break
+				}
+				serverRule, ok := serverRuleStore.(*k8s.CiliumNetworkPolicy)
+				if !ok {
+					log.Warningf("Received unknown object %+v, expected a CiliumNetworkPolicy object", obj)
+					return
+				}
+				if serverRule.Metadata.UID != rule.Metadata.UID {
+					// Although the policy was found this means it was deleted,
+					// and re-added with the same name.
+					log.Debugf("k8s: rule %s/%s changed while updating node status, stopping retry", ns, name)
+					break
+				}
+				serverRule.SetPolicyStatus(nodeaddress.GetName(), cnpns)
+				_, err = crdClient.Update(serverRule)
```

Comment:
Doesn't `Update()` give you the latest revision of the resource when it fails? You can use it instead of hoping for a race free get & update.

Reply:
>Doesn't Update() give you the latest revision of the resource when it fails? You can use it instead of hoping for a race free get & update.

It doesn't and it took me a while to figure out that. How can we still get a race with the get & update? I was hopping this would prevent it:
```
				if serverRule.Metadata.UID != rule.Metadata.UID {
					// Although the policy was found this means it was deleted,
					// and re-added with the same name.
					log.Debugf("k8s: rule %s/%s changed while updating node status, stopping retry", ns, name)
					break
				}
```
but I can modify it to this as well:
```
				if serverRule.Metadata.UID != rule.Metadata.UID ||
                                       !serverRule.SpecEquals(rule)   {
					// Although the policy was found this means it was deleted,
					// and re-added with the same name.
					log.Debugf("k8s: rule %s/%s changed while updating node status, stopping retry", ns, name)
					break
				}
```

Reply:
The code is fine but it looked inefficient. It's unexpected that `Update()` does not provide the latest version.